jax/interpreters/batching.py

```diff
@@ -122,27 +122,34 @@ def lift(self, val):
   def sublift(self, val):
     return BatchTracer(self, val.val, val.batch_dim)
 
-  def get_primitive_batcher(self, primitive):
-    if primitive in initial_style_batchers:
-      return partial(initial_style_batchers[primitive],
-                    axis_name=self.axis_name,
-                    main_type=self.main.trace_type)
-    try:
+  def get_primitive_batcher(self, primitive, frame):
+    if primitive in primitive_batchers:
       return primitive_batchers[primitive]
-    except KeyError as err:
-      msg = "Batching rule for '{}' not implemented"
-      raise NotImplementedError(msg.format(primitive)) from err
+    elif primitive in collective_rules:
+      return partial(collective_rules[primitive], frame)
+    msg = "Batching rule for '{}' not implemented"
+    raise NotImplementedError(msg.format(primitive))
+
+  def get_frame(self, vals, dims) -> core.AxisEnvFrame:
+    if self.axis_name is core.no_axis_name:
+      # If axis name is `no_axis_name` we can't find it via `core.axis_name` so we
+      # reconstruct it from the information we have available
+      axis_sizes = {x.shape[d] for x, d in zip(vals, dims) if d is not not_mapped}
+      assert len(axis_sizes) == 1
+      axis_size, = axis_sizes
+      return core.AxisEnvFrame(self.axis_name, axis_size, self.main)
+    return core.axis_frame(self.axis_name)
 
   def process_primitive(self, primitive, tracers, params):
     vals_in, dims_in = unzip2((t.val, t.batch_dim) for t in tracers)
+    frame = self.get_frame(vals_in, dims_in)
     if (primitive in collective_rules and
           _main_trace_for_axis_names(self.main, core.used_axis_names(primitive, params))):
-      frame = core.axis_frame(self.axis_name)
       val_out, dim_out = collective_rules[primitive](frame, vals_in, dims_in, **params)
     elif all(bdim is not_mapped for bdim in dims_in):
       return primitive.bind(*vals_in, **params)
     else:
-      batched_primitive = self.get_primitive_batcher(primitive)
+      batched_primitive = self.get_primitive_batcher(primitive, frame)
       val_out, dim_out = batched_primitive(vals_in, dims_in, **params)
```

Comment:
Instead of adding `frame` to `get_primitive_batcher`, couldn't we just remove the `_main_trace_for_axis_name` condition from the first `if` case?

Reply:
I don't think so. There are axis primitives (e.g. psum I believe) that have a standard primitive batcher in addition to an axis primitive rule. In this case, I think the logic needs to follow:
1. If it's not an axis primitive, use the standard primitive batcher
2. If it is an axis primitive (collective or initial style):
  * If the axis name matches, use the axis primitive batcher (overlapping case)
  * If the axis name does not match:
     * If there is a standard batcher, use it (collectives case)
     * Otherwise, use the axis primitive batcher (initial style primitives case)

I've made this more explicit in `process_primitive`, PTAL.